### ISSUE TITLE
Some order pre-test seem not to be as specific as would be appropirate

### ISSUE THREAD

**Issue by LynxAbraxas** [civctp2/civctp2#179]
In relation to https://github.com/civctp2/civctp2/pull/176#issuecomment-514636722, this issue is for tracking the problem that apparently some order pre-test seem not to be as specific as would be appropriate, see e.g. https://github.com/civctp2/civctp2/issues/169#issuecomment-514366469.
This seems to lead to execution of orders that then fail due to the tests made in the function trying to execute that order. This is non-ideal since the AI logic builts on the pre-tests results when deciding what to order a unit to do. If the pre-tests lead to a successful prediction, the AI orders an army to do so which then fails and in the end uses up the army move points and its abilities.
This might also be the source for the irritating cashier sound (that lets one associate success) in case such an order does not succeed, possibly the source for #54.

**Comment by LynxAbraxas** [comment]
A possible source for the odd sound could be this logic:
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/gs/gameobj/ArmyData.cpp#L9733

**Comment by LynxAbraxas** [comment]
See also https://github.com/civctp2/civctp2/pull/176#issuecomment-514992966.

**Comment by LynxAbraxas** [comment]
@MartinGuehmann  If lack of gold is the actual source of the problem, would it be appropriate to put some kind of check like
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/gs/gameobj/ArmyData.cpp#L10468
into here
https://github.com/civctp2/civctp2/blob/2c9b619e1bb4fe4e3fab1b86dddf521482e930d7/ctp2_code/ai/strategy/goals/Goal.cpp#L2717-L2976

**Comment by LynxAbraxas** [comment]
The sound seems to come from
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/gs/gameobj/ArmyData.cpp#L9455-L9474
in particular
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/gs/gameobj/ArmyData.cpp#L9468

**Comment by MartinGuehmann** [comment]
> @MartinGuehmann If lack of gold is the actual source of the problem, would it be appropriate to put some kind of check like
> https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/gs/gameobj/ArmyData.cpp#L10468
> 
> 
> into here
> https://github.com/civctp2/civctp2/blob/2c9b619e1bb4fe4e3fab1b86dddf521482e930d7/ctp2_code/ai/strategy/goals/Goal.cpp#L2717-L2976

No, just because you cannot pay the attack, doesn't mean that the target is immune against the attack. The gold thing should go into Goal::IsCurrentlyUnavailable(). However, as far as I can see that has been already handled.

**Comment by LynxAbraxas** [comment]
> No, just because you cannot pay the attack, doesn't mean that the target is immune against the attack. The gold thing should go into Goal::IsCurrentlyUnavailable(). However, as far as I can see that has been already handled.

True and it is handled here
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/ai/strategy/goals/Goal.cpp#L2987-L2993
and that handling gets called but still some logic (probably afterwards) somehow leads to the result that the AI tries to execute that goal despite insufficient gold.

**Comment by LynxAbraxas** [comment]
I'm not feeling confident in understanding the whole goal logic but could it be that this
https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/ai/strategy/goals/Goal.cpp#L2599
should be `return IsComplete() && !IsCurrentlyUnavailable();`? 
It's the only place where `IsCurrentlyUnavailable` is used.

**Comment by MartinGuehmann** [comment]
> I'm not feeling confident in understanding the whole goal logic but could it be that this
> https://github.com/civctp2/civctp2/blob/82a15c92ee2b85adf01e56f89e334a15e8926dbb/ctp2_code/ai/strategy/goals/Goal.cpp#L2599
> 
> 
> should be `return IsComplete() && !IsCurrentlyUnavailable();`?
> It's the only place where `IsCurrentlyUnavailable` is used.

No, the difference between IsComplete() and IsTotallyComplete() is that the AI should not send units to goals that are complete, because these goals won't be valid in the next turns again. However, the AI should still send units to goals that are just not executable, because of lack of gold. It may need quite some time to send a unit oversea. And just aborting and resuming, because you have temporarily no gold makes the AI behavior quite unstable.

At least that is the idea behind it. If you should call these things then IsComplete() and IsTotallyComplete() is another issue.

**Comment by LynxAbraxas** [comment]
Many thanks @MartinGuehmann for that explantation, with that I understand the logic a bit better.

> At least that is the idea behind it. If you should call these things then IsComplete() and IsTotallyComplete() is another issue.

Yes, the naming was irritating me but a rename just for that might be a bit of an overkill.

So where do you think could the problem lie that the AI still tries to execute some orders in the very same turn where there is not enough money left?

**Comment by MartinGuehmann** [comment]
> Yes, the naming was irritating me but a rename just for that might be a bit of an overkill.

Why is renaming them an overkill? The names are supposed to give us some idea what these functions are about. In fact I think I already removed some underscores from them. Actually, MSVC studio has a function for exactly doing this. Even so I have never used it, probably the previous versions of MSVC hadn't it. But the idea is MSVC replaces all the instances of that function automatically with the new name.
 
> So where do you think could the problem lie that the AI still tries to execute some orders in the very same turn where there is not enough money left?

I guess that it is exactly what you have brought up. The question is whether it is a problem? Does it really cost move points, which of course shouldn't since you cannot execute the order. And even if it does, shouldn't then the unit stay where it is so that it can try next turn again. In that case it doesn't matter whether it wastes move points.

**Comment by LynxAbraxas** [comment]
If the AI moves the Lawyer to a city with the intention to use it there in another turn, that is great.
Could it be that the goal is evalutated before the AI lacks gold, i.e. before gold is removed for the first removal of a Franchise?

>  The question is whether it is a problem?

If the effect does not weaken the AI (by losing move points or pursuing a goal invain) then probably not.

**Comment by MartinGuehmann** [comment]
> If the AI moves the Lawyer to a city with the intention to use it there in another turn, that is great.
> Could it be that the goal is evalutated before the AI lacks gold, i.e. before gold is removed for the first removal of a Franchise?

Well, that is also true. Moving, loading transporters, attacking, suing franchises are all done from events. These events are processed asynchronously. Therefore, you have the process match cycles in each round you schedule the events. For instance in the first cycle you load a transporter and in the next you can then move the transporter.

This system allows you to extend the code by slic or actually, within c++ code by adding modules that fire before or after the main code.

But since the events are executed asynchronously that is a bit error prone, and so I used some time to debug this. So basically you give all the orders to the armies in one cycle then the orders are executed and then we go to the next cycle.

Of course at that point we don't know how much gold is left after that cycle. And that would be quite a hassle to track. We may give a certain amount of orders that cost so much gold if executed, but we don't know how many units are at the target. So that is a bit impractical to record, so I just prefer that the units try and that's it.

> 
> > The question is whether it is a problem?
> 
> If the effect does not weaken the AI (by losing move points or pursuing a goal invain) then probably not.

Then we just have to check, whether the units lose move points. And fix it if it does.

**Comment by LynxAbraxas** [comment]
OK that makes some things clearer, thank for that explanation.

> Then we just have to check, whether the units lose move points. And fix it if it does.

Shall I close the issue until we observe this or leave it open any way?

**Comment by MartinGuehmann** [comment]
> OK that makes some things clearer, thank for that explanation.
> 
> > Then we just have to check, whether the units lose move points. And fix it if it does.
> 
> Shall I close the issue until we observe this or leave it open any way?

I think it should be enough to test it as a human player, or with an AI player on the screen.
